### PR TITLE
Add Windows Jumplist plugin

### PR DESCRIPTION
**Pull request by Zawadidone**
```
target-query -t TARGET -f jumplist  --limit 1 | rdump -L
[reading from stdin]
--[ RECORD 1 ]--
          hostname = WINDEV2401EVAL
            domain = None
              type = customDestinations
    application_id = 590aee7bdd69b59b
  application_name = Powershell Windows 10
          lnk_path = C:\Users\User\AppData\Roaming\Microsoft\Windows\Recent\CustomDestinations\590aee7bdd69b59b.customDestinations-ms
          lnk_name = None
         lnk_mtime = 2024-03-21 17:16:41.027615+00:00
         lnk_atime = 2024-03-21 17:16:40.997318+00:00
         lnk_ctime = 2024-03-21 17:16:29.070940+00:00
  lnk_relativepath = None
       lnk_workdir = None
     lnk_arguments = None
  lnk_iconlocation = %windir%\System32\WindowsPowerShell\v1.0\powershell.exe
   local_base_path = C:\Users\User\AppData\Roaming\Microsoft\Windows\Start Menu\Programs\Windows PowerShell\Windows PowerShell.lnk
common_path_suffix = 
      lnk_net_name = None
   lnk_device_name = None
     lnk_full_path = C:\Users\User\AppData\Roaming\Microsoft\Windows\Start Menu\Programs\Windows PowerShell\Windows PowerShell.lnk
        machine_id = windev2401eval
      target_mtime = 2022-05-07 05:22:32.504461+00:00
      target_atime = 2024-03-21 17:16:28.045404+00:00
      target_ctime = 2024-03-21 17:16:20.202059+00:00
          username = User
           user_id = S-1-5-21-147454635-2304731113-4176578439-1000
        user_group = None
         user_home = C:\Users\User
[...]
```

### CODE REVIEW

**Review by Horofic — CHANGES_REQUESTED** [review]

File: dissect/target/helpers/utils.py [hunk 1]
```diff
@@ -13,6 +13,17 @@
 log = logging.getLogger(__name__)
 
 
+def findall(buf, needle):
```
[the remark this answers is not in the document]
Comment (Zawadidone) [reply dissect/target/helpers/utils.py:16]: Fixed

File: dissect/target/plugins/os/windows/jumplist.py [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+                    type=type,
+                    application_name=application_name,
+                    application_id=application_id,
+                    **lnk._asdict(),
+                    _user=user,
+                    _target=self.target,
+                )
+
+    @export(record=JumpListRecord)
+    def jumplist(self) -> Iterator[JumpListRecord]:
```
[the remark this answers is not in the document]
Comment (Zawadidone) [reply dissect/target/plugins/os/windows/jumplist.py:238]: Is it a known issue that namespace plugins do not work?

`AttributeError: 'JumpListPlugin' object has no attribute 'SUBPLUGINS'`

File: dissect/target/plugins/os/windows/jumplist.py [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+
+struct header_end_0 {
+    uint16  name_length;
+    wchar   name[name_length];
+    int     number_of_entries;
+}
+
+struct entry_header {
+    char guid[16];
+}
```
[the remark this answers is not in the document]
Comment (Zawadidone) [reply dissect/target/plugins/os/windows/jumplist.py:55]: Fixed

File: dissect/target/plugins/os/windows/jumplist.py [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+            raise UnsupportedPluginError("No Jump List files found")
+
+    def name(self, name: str) -> tuple[Any, str, str]:
+        application_id, application_type = name.split(".")
+        application_type = application_type.split("-")[0]
+        application_name = APPLICATION_IDENTIFIERS.get(application_id)
+
+        return application_name, application_id, application_type
+
+    def custom_destination(self) -> Iterator[JumpListRecord]:
```
[the remark this answers is not in the document]
Comment (Zawadidone) [reply dissect/target/plugins/os/windows/jumplist.py:158]: Fixed

File: dissect/target/plugins/os/windows/jumplist.py [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+                yield JumpListRecord(
+                    type=application_type,
+                    application_name=application_name,
+                    application_id=application_id,
+                    **lnk._asdict(),
+                    _user=user,
+                    _target=self.target,
+                )
+
+    def automatic_destination(self):
```
[the remark this answers is not in the document]
Comment (Zawadidone) [reply dissect/target/plugins/os/windows/jumplist.py]: Fixed

File: tests/test_plugins_os_windows_jumplist.py
[the remark this answers is not in the document]
Comment (Zawadidone) [reply tests/test_plugins_os_windows_jumplist.py:1]: Fixed

File: dissect/target/plugins/os/windows/jumplist.py [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+
+    def custom_destination(self) -> Iterator[JumpListRecord]:
+        for destination, user in self.custom_destinations:
+            fh = destination.open("rb")
+
+            try:
+                custom_destination = CustomerDestinationFile(fh)
+            except EOFError:
+                continue
+            except NotImplementedError:
```
[the remark this answers is not in the document]
Comment (Zawadidone) [reply dissect/target/plugins/os/windows/jumplist.py:166]: Fixed

File: dissect/target/plugins/os/windows/jumplist.py [hunk 7]
```diff
@@ -0,0 +1,255 @@
...
+                    custom_destination.header.value_type,
+                    destination,
+                )
+                continue
+            except Exception as e:
+                self.target.log.warning("Failed to parse CustomDestination header: %s", destination)
+                self.target.log.debug("", exc_info=e)
+                continue
+
+            if not custom_destination.MAGIC_FOOTER == custom_destination.magic:
```
[the remark this answers is not in the document]
Comment (Zawadidone) [reply dissect/target/plugins/os/windows/jumplist.py:178]: Fixed

File: dissect/target/plugins/os/windows/jumplist.py [hunk 8]
```diff
@@ -0,0 +1,255 @@
...
+            for destination in user_details.home_path.glob(
+                "AppData/Roaming/Microsoft/Windows/Recent/AutomaticDestinations/*.automaticDestinations-ms"
+            ):
+                self.automatic_destinations.append([destination, user_details.user])
+
+    def check_compatible(self) -> None:
+        if not any([self.automatic_destinations, self.custom_destinations]):
+            raise UnsupportedPluginError("No Jump List files found")
+
+    def name(self, name: str) -> tuple[Any, str, str]:
```
[the remark this answers is not in the document]
Comment (Zawadidone) [reply dissect/target/plugins/os/windows/jumplist.py]: Fixed

File: dissect/target/plugins/os/windows/jumplist.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Horofic) [reply dissect/target/plugins/os/windows/jumplist.py]: An issue with nested namespaced plugin was reported here https://github.com/fox-it/dissect.target/issues/758 and fixed in this PR https://github.com/fox-it/dissect.target/pull/763. You could try and check-out this PR to see if you issue persists!

[the remark this answers is not in the document]
Comment (Zawadidone) [reply dissect/target/plugins/os/windows/jumplist.py]: The issue persists.

**Review by Horofic — CHANGES_REQUESTED** [review]

**Comment by Horofic** [comment]
@Zawadidone I also committed some suggestions directly in your PR, please check if you see anything wrong :).

File: dissect/target/plugins/os/windows/jumplist.py [hunk 9]
```diff
@@ -0,0 +1,253 @@
...
+                    type=application_type,
+                    application_name=application_name,
+                    application_id=application_id,
+                    **lnk._asdict(),
+                    _user=user,
+                    _target=self.target,
+                )
+
+    @export(record=JumpListRecord)
+    def automatic_destination(self) -> Iterator[JumpListRecord]:
```
[the remark this answers is not in the document]
Comment (Zawadidone) [reply dissect/target/plugins/os/windows/jumplist.py:231]: Fixed

File: dissect/target/plugins/os/windows/jumplist.py [hunk 10]
```diff
@@ -0,0 +1,253 @@
...
+                "AppData/Roaming/Microsoft/Windows/Recent/AutomaticDestinations/*.automaticDestinations-ms"
+            ):
+                self.automatic_destinations.append([destination, user_details.user])
+
+    def check_compatible(self) -> None:
+        if not any([self.automatic_destinations, self.custom_destinations]):
+            raise UnsupportedPluginError("No Jump List files found")
+
+    @export(record=JumpListRecord)
+    def custom_destination(self) -> Iterator[JumpListRecord]:
```
[the remark this answers is not in the document]
Comment (Zawadidone) [reply dissect/target/plugins/os/windows/jumplist.py:197]: Fixed

**Comment by Zawadidone** [comment]
[`f9fb309` (#669)](https://github.com/fox-it/dissect.target/pull/669/commits/f9fb30980789cf7562faa7c5abcdfbd6d8805f0b) includes is a change based on [`6464029` (#669)](https://github.com/fox-it/dissect.target/pull/669/commits/646402979f555eb0fcaeee97638046f0264fffd1), if this is not preferred I can remove the commit from the branch.

**Comment by Horofic** [comment]
> [`f9fb309` (#669)](https://github.com/fox-it/dissect.target/pull/669/commits/f9fb30980789cf7562faa7c5abcdfbd6d8805f0b) is based on [`6464029` (#669)](https://github.com/fox-it/dissect.target/pull/669/commits/646402979f555eb0fcaeee97638046f0264fffd1), if this is not preferred I can remove the commit from the branch.

[6464029 (#669)](https://github.com/fox-it/dissect.target/pull/669/commits/646402979f555eb0fcaeee97638046f0264fffd1) is fine! That actually was the first approach I took :)

File: dissect/target/plugins/os/windows/jumplist.py [hunk 11]
```diff
@@ -0,0 +1,286 @@
+import io
+import logging
+from struct import error as StructError
+from typing import BinaryIO, Callable, Iterator
```
Comment (Horofic) [inline dissect/target/plugins/os/windows/jumplist.py:4]:
```suggestion
from __future__ import annotations

import io
import logging
from struct import error as StructError
from typing import BinaryIO, Iterator
```

File: dissect/target/plugins/os/windows/jumplist.py [hunk 12]
```diff
@@ -0,0 +1,286 @@
...
+            lnk_net_name (string): Specifies a server share path; for example, "\\\\server\\share".
+            lnk_device_name (string): Specifies a device; for example, the drive letter "D:"
+            machine_id (string): The NetBIOS name of the machine where the linked file was last known to reside.
+            target_mtime (datetime): Modification time of the target (linked) file.
+            target_atime (datetime): Access time of the target (linked) file.
+            target_ctime (datetime): Creation time of the target (linked) file.
+        """
+        yield from self._generate_records(self.automatic_destinations, AutomaticDestinationFile)
+
+    def _generate_records(self, destinations: list, destination_file: Callable) -> Iterator[JumpListRecord]:
```
Comment (Horofic) [inline dissect/target/plugins/os/windows/jumplist.py:264]:
```suggestion
    def _generate_records(
        self, destinations: list, destination_file: AutomaticDestinationFile | CustomDestinationFile
    ) -> Iterator[JumpListRecord]
```

**Review by Horofic — CHANGES_REQUESTED** [review]

File: dissect/target/plugins/os/windows/jumplist.py (hunk 12 above)
Comment (Zawadidone) [reply dissect/target/plugins/os/windows/jumplist.py]: Fixed

File: dissect/target/plugins/os/windows/jumplist.py (hunk 11 above)
Comment (Zawadidone) [reply dissect/target/plugins/os/windows/jumplist.py]: Fixed

**Review by Horofic — APPROVED** [review]

**Pull request merged**